linera-views/src/backends/lru_caching.rs

```diff
@@ -22,25 +36,259 @@ use crate::memory::MemoryStore;
+        }
+    }
+
+    fn size(&self) -> usize {
+        match self {
+            ValueCacheEntry::DoesNotExist => 0,
+            ValueCacheEntry::Exists => 0,
+            ValueCacheEntry::Value(vec) => vec.len(),
+        }
+    }
+}
+
+enum FindCacheEntry {
+    Keys(Vec<Vec<u8>>),
+    KeyValues(Vec<(Vec<u8>, Vec<u8>)>),
+}
+
+impl FindCacheEntry {
+    fn get_find_keys(&self, key_prefix: &[u8]) -> Vec<Vec<u8>> {
+        let delta = key_prefix.len();
+        let mut keys = Vec::new();
+        match self {
+            FindCacheEntry::Keys(vec) => {
+                for key in vec {
+                    if key.starts_with(key_prefix) {
+                        keys.push(key[delta..].to_vec());
+                    }
+                }
+            }
+            FindCacheEntry::KeyValues(vec) => {
+                for (key, _) in vec {
+                    if key.starts_with(key_prefix) {
+                        keys.push(key[delta..].to_vec());
+                    }
+                }
+            }
+        }
+        keys
+    }
+
+    fn get_find_key_values(&self, key_prefix: &[u8]) -> Option<Vec<(Vec<u8>, Vec<u8>)>> {
+        match self {
+            FindCacheEntry::Keys(_) => None,
+            FindCacheEntry::KeyValues(vec) => {
+                let delta = key_prefix.len();
+                let mut key_values = Vec::new();
+                for (key, value) in vec {
+                    if key.starts_with(key_prefix) {
+                        let key_value = (key[delta..].to_vec(), value.clone());
+                        key_values.push(key_value);
+                    }
+                }
+                Some(key_values)
+            }
+        }
+    }
+
+    fn get_contains_key(&self, key: &[u8]) -> bool {
+        let key = key.to_vec();
+        match self {
+            FindCacheEntry::Keys(vec) => {
+                let result = vec.binary_search_by(|probe| probe.cmp(&key));
+                result.is_ok()
+            }
+            FindCacheEntry::KeyValues(vec) => {
+                let result = vec.binary_search_by(|(probe, _)| probe.cmp(&key));
+                result.is_ok()
+            }
+        }
+    }
+
+    fn get_read_value(&self, key: &[u8]) -> Option<Option<Vec<u8>>> {
+        let key = key.to_vec();
+        match self {
+            FindCacheEntry::Keys(_vec) => None,
+            FindCacheEntry::KeyValues(vec) => {
+                let result = vec.binary_search_by(|(probe, _)| probe.cmp(&key));
+                match result {
+                    Ok(pos) => Some(Some(vec[pos].1.clone())),
+                    Err(_) => Some(None),
+                }
+            }
+        }
+    }
+
+    fn update_cache_entry(&mut self, key: &[u8], new_value: Option<Vec<u8>>) {
+        let key = key.to_vec();
+        match self {
+            FindCacheEntry::Keys(vec) => {
+                let result = vec.binary_search_by(|probe| probe.cmp(&key));
```

Comment:
I used the `binary_seach*` functions for the first time (it was pretty intensive last week).
I think the use of `BTreeSet` / `BTreeMap` is not warranted because the cost of those data structures is larger than the one of vec.